examples/asr_training/rnnt_lightning.py

```diff
@@ -0,0 +1,535 @@
+
+from fairseq.data import Dictionary
+import sentencepiece as spm
+
+import torch
+import torchaudio
+import torchaudio.functional as F
+from torchaudio.prototype.rnnt import emformer_rnnt_base
+from torchaudio.prototype.rnnt_decoder import Hypothesis, RNNTBeamSearch
+from pytorch_lightning import LightningModule, Trainer
+from pytorch_lightning.callbacks import ModelCheckpoint
+
+
+logger = logging.getLogger()
+
+Batch = namedtuple(
+    "Batch", ["features", "feature_lengths", "targets", "target_lengths"]
+)
+
+
+_decibel = 2 * 20 * math.log10(torch.iinfo(torch.int16).max)
+_gain = pow(10, 0.05 * _decibel)
+
+_spectrogram_transform = torchaudio.transforms.MelSpectrogram(
+    sample_rate=16000, n_fft=400, n_mels=80, hop_length=160
+)
+
+
+def _batch_by_token_count(idx_target_lengths, token_limit):
+    batches = []
+    current_batch = []
+    current_token_count = 0
+    for idx, target_length in idx_target_lengths:
+        if current_token_count + target_length > token_limit:
+            batches.append(current_batch)
+            current_batch = [idx]
+            current_token_count = target_length
+        else:
+            current_batch.append(idx)
+            current_token_count += target_length
+
+    if current_batch:
+        batches.append(current_batch)
+
+    return batches
+
+
+class CustomDataset(torch.utils.data.Dataset):
+    r"""Sort samples by target length and batch to max token count."""
+
+    def __init__(self, base_dataset, max_token_limit):
+        super().__init__()
+        self.base_dataset = base_dataset
+
+        fileid_to_target_length = {}
+        idx_target_lengths = [
+            (idx, self._target_length(fileid, fileid_to_target_length))
+            for idx, fileid in enumerate(self.base_dataset._walker)
+        ]
+
+        assert len(idx_target_lengths) > 0
+
+        idx_target_lengths = sorted(
+            idx_target_lengths, key=lambda x: x[1], reverse=True
+        )
+
+        assert max_token_limit >= idx_target_lengths[0][1]
+
+        self.batches = _batch_by_token_count(idx_target_lengths, max_token_limit)
+
+    def _target_length(self, fileid, fileid_to_target_length):
+        if fileid not in fileid_to_target_length:
+            speaker_id, chapter_id, _ = fileid.split("-")
+
+            file_text = speaker_id + "-" + chapter_id + self.base_dataset._ext_txt
+            file_text = os.path.join(
+                self.base_dataset._path, speaker_id, chapter_id, file_text
+            )
+
+            with open(file_text) as ft:
+                for line in ft:
+                    fileid_text, transcript = line.strip().split(" ", 1)
+                    fileid_to_target_length[fileid_text] = len(transcript)
+
+        return fileid_to_target_length[fileid]
+
+    def __getitem__(self, idx):
+        return [self.base_dataset[subidx] for subidx in self.batches[idx]]
+
+    def __len__(self):
+        return len(self.batches)
+
+
+class TimeMasking(torchaudio.transforms._AxisMasking):
```

Comment:
I found it in transforms but it's not implemented yet. If it's a standard process for augmentation we can move it to transforms after this PR.

Reply:
yeah i can add the `min_mask_p` parameter to the existing `TimeMasking` in transforms later